Adding Gaussian Mixture Model

krstopro opened the pull request:
Added Gaussian Mixture Model implementation, similar to the one in scikit-learn Python module. Currently it supports only full covariances and uses k-means for the initialisation. Other covariance types (diagonal, tied, spherical) and initialisation methods (random, k-means++) can easily be added.

msluszniak reviewed (approved):
Thank you for your contribution! The PR LGTM :)

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+  @moduledoc """
+  Gaussian Mixture Model.
+
+  Gaussian Mixture model is a probabilistic model that assumes every data point is generated
```
```suggestion
  Gaussian Mixture Model is a probabilistic model that assumes every data point is generated
```

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+  Gaussian Mixture Model.
+
+  Gaussian Mixture model is a probabilistic model that assumes every data point is generated
+  by randomly choosing one of the several fixed Gaussian distributions and then sampling from it.
```
"randomly choosing" might be a loaded expression here. I got stuck on thinking about if this is random uniform, normal or whatever else method of choosing.

```suggestion
  b choosing one of several fixed Gaussian distributions and then sampling from it.
```

krstopro replied:
Yes, it is a multinomial random distribution. Basically GMM can be viewed as a generative model in the following way:
- First sample one out of K Gaussian distributions from multinomial random distribution with parameters w_1, ..., w_K (w_k is the probability of sampling k-th Gaussian; this is the `weights` field of the model).
- Then sample from that Gaussian distribution.

krstopro replied:
Agreed about the suggestion though.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+  expectation of the Gaussian assignment for each data point x and the M-step which updates the
+  parameters to maximize the expectations found in E-step. While every iteration of the algorithm
+  is guaranteed to improve the log-likelihood, the final result depends on the initial values of
+  the parameters and the entire procedure should be repeated several times.
```
From reading without looking at the code beforehand:

Does this mean that we need to run the algorithm multiple times, or does the algorithm itself repeat this procedure?

krstopro replied:
The algorithm itself repeats the procedure `num_runs` amount of times, each time with different initial parameters. Similarly as it was already done in k-means implementation.

krstopro replied:
I am not sure if I should change anything here. Thoughts?

msluszniak replied:
For me, the description is clear. @polvalente do you have any suggestions?

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+    num_gaussians: [
+      required: true,
+      type: :pos_integer,
+      doc: "The number of Gaussian components in the mixture."
```
```suggestion
      doc: "The number of Gaussian model components in the mixture."
```

krstopro replied:
The Gaussian here refers to the Gaussian distribution. Perhaps writing "The number of Gaussian distributions in the mixture" would be better.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+    num_runs: [
+      type: :pos_integer,
+      default: 1,
+      doc: "The number of initializations to perform."
```
Not sure if "initializations" is the best word because I assume we're talking about running the full procedure and not just initializing it, right?

krstopro replied:
Yes, running the full procedure but with different initial parameters. Indeed, word initializations might be misleading here.

krstopro replied:
I guess "The number of times to initialize parameters and run the entire EM algorithm." might be a better description here.

msluszniak replied:
Agreed. This description will be maximally informative.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+      default: 1.0e-3,
+      doc: "The convergence threshold."
+    ],
+    reg_covar: [
```
```suggestion
    covariance_regularization_eps: [
```

krstopro replied:
Agreed.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+      default: 1.0e-6,
+      doc: """
+      The non-negative number that is added to each element of the diagonal
+      of the covariance matrix to ensure it is positive.
```
```suggestion
      of the covariance matrix to ensure it is positive. Usually a small number.
```

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+                reg_covar
+              }
+            },
+            run <= num_runs do
```
Since there's no early stopping, we might be able to apply vectorization to get rid of this outer while loop

msluszniak replied:
I think that in the case of k-means, vectorization is more beneficial. You can simply reduce `num_runs` if the memory is the problem. But you cannot apply vectorization programmatically if you use two nested whiles. Nowadays, computation time is much more important than space limitation. You might benchmark two versions on the example that can be loaded into memory and test how much faster the vectorized version is. If the change is insignificant for GMM, we should stay with a memory-efficient solution.

msluszniak replied:
And it's always easier to change vectorized version into 'while', so if people had a memory problem on k-means, then we would use a memory-efficient solution.

msluszniak replied:
And in Sci-kit they cannot run k-means on GPU, so they have no clear advantage of implementing vectorization which is not obvious to understand at first glance, harder to implement (because of broadcasting), memory consuming etc.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+  end
+
+  defnp e_step(x, weights, means, precisions_cholesky) do
+    {log_prob_norm, log_resp} = estimate_log_prob_resp(x, weights, means, precisions_cholesky)
```
what does _resp mean here?

krstopro replied:
`resp` is short for responsibility or responsibilities. `resp[n][k]` measure how much is the k-th Gaussian responsible for generating the n-th observation. Perhaps using `responsibilities` instead of `resp` might be a better idea.

krstopro replied:
Not sure about this as well. Thoughts?

msluszniak replied:
If it wasn't intuitive for Paulo, then I think that it is worth changing it to `responsibilities`.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+
+  defnp estimate_gaussian_parameters(x, resp, reg_covar) do
+    nk = Nx.sum(resp, axes: [0])
+    means = Nx.dot(Nx.transpose(resp), x) / nk
```
it might be more numerically stable to do:

```suggestion
    means = Nx.dot(Nx.transpose(resp) / nk, x / nk)
```

polvalente replied:
Also, if `resp` can be complex, you might want Nx.LinAlg.adjoint instead

krstopro replied:
`resp` are probabilities, so `Nx.transpose` will suffice. Should it be `Nx.sqrt(nk)` in the second expression? Is this really more numerically stable?

polvalente replied:
If resp are probabilities then it won't make much difference, but you can just divide the other term by nk instead.

My concern was that for 2 unbounded tensors, the dot product could blow up a bit, and the division tames that to some degree.

Since one of those is bounded between 0 and 1 in each element, this isn't that much of a concern

krstopro replied:
Indeed, dot product can explode, but here it should not be a problem. Rewriting it as
` means = Nx.dot(Nx.transpose(resp), x / nk)` might be a good idea though.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+    type = Nx.type(x)
+    {num_gaussians, num_features} = Nx.shape(means)
+
+    {covariances, _} =
+      while {
+              covariances =
+                Nx.tensor(0.0, type: type)
+                |> Nx.broadcast({num_gaussians, num_features, num_features}),
+              {k = 0, x, resp, nk, means, reg_covar}
+            },
+            k < num_gaussians do
+        diff = x - means[k]
+        covariance = Nx.dot(resp[[.., k]] * Nx.transpose(diff), diff) / nk[k]
+        covariance = Nx.put_diagonal(covariance, Nx.take_diagonal(covariance) + reg_covar)
+
+        {
+          Nx.put_slice(covariances, [k, 0, 0], Nx.new_axis(covariance, 0)),
+          {k + 1, x, resp, nk, means, reg_covar}
+        }
+      end
```
This while also reads like something that can benefit from vectorization

polvalente replied:
Also, we can leave usage of vectorization as TO-DO comments instead, unless you really want to tackle this. It might even be the case where we merge this and you tackle it afterwards. Thoughts?

krstopro replied:
Indeed, this might be vectorizable. I would need to think about lines 255 and 256 though as it's not that straightforward.
I would say let's merge it like this and I can continue working on it (I intend to continue contributing to `Nx` anyway).

polvalente replied:
Sure! To be clear, vectorization is not a blocker at all for this PR

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+    {num_gaussians, num_features, _} = Nx.shape(covariances)
+
+    {precisions_cholesky, _} =
+      while {
```
Same thing about vectorization here

krstopro replied:
Yeah, this one might be easier as both `Nx.LinAlg.cholesky` and `Nx.LinAlg.triangular_solve` work with batches of matrices.

polvalente replied:
You don't have to think much about this. Nx main has Nx.vectorize/devectorize/revectorize which will handle batching for you

krstopro replied:
One thing was strange to me. Variable `type` is not declared in the `while` loop and it still compiles. Is this intended?

polvalente replied:
In fact, I'm not sure it _would_ compile if you passed type as an argument :)

type is a tuple of {atom, number} used only at compile time. Passing as argument would force its traversal as an Nx.Container and that would break on the Atom.

It's ok anyway because the type is just used at compile-time

josevalim replied:
Nx only checks the arguments between def <-> defn boundaries.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+    log_det = compute_log_det_cholesky(precisions_cholesky)
+
+    {log_prob, _} =
+      while {
```
vectorization too

krstopro replied:
Yeah, also doesn't seem hard to vectorize.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+      iex> x = Nx.tensor([[1, 2], [1, 4], [1, 0], [10, 2], [10, 4], [10, 0]])
+      iex> model = Scholar.Cluster.GaussianMixture.fit(x, num_gaussians: 2, key: key)
+      iex> Scholar.Cluster.GaussianMixture.predict(model, Nx.tensor([[8, 1], [2, 3]]))
+      Nx.tensor(
+        [1, 0]
+      )
```
I don't think this will pass on CI

krstopro replied:
Hmmm why exactly? `mix test` also checks the docstrings and runs the code inside.
It's similar to the k-means implementation:
https://github.com/elixir-nx/scholar/blob/main/lib/scholar/cluster/k_means.ex#L297-L299

josevalim replied:
Yup, this is a doctest feature :)

polvalente replied:
TIL :)

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+      iex> x = Nx.tensor([[1, 2], [1, 4], [1, 0], [10, 2], [10, 4], [10, 0]])
+      iex> model = Scholar.Cluster.GaussianMixture.fit(x, num_gaussians: 2, key: key)
+      iex> Scholar.Cluster.GaussianMixture.predict_prob(model, Nx.tensor([[8, 1], [2, 3]]))
+      Nx.tensor(
+        [
+          [0.0, 1.0],
+          [1.0, 0.0]
+        ]
+      )
```
This as well

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,407 @@
+            params = m_step(x, log_resp, reg_covar)
+            lower_bound = log_prob_norm
+            change = lower_bound - prev_lower_bound
+            converged = Nx.abs(change) < tol
```
```suggestion
            converged = iter > 1 and Nx.abs(change) < tol
```

krstopro replied:
Why is `iter > 1` necessary here?

polvalente replied:
My previous version of the comment had an explanation haha

I think it would be nice to avoid relying on comparison with neg infinity (the initial value) in the first iteration

krstopro replied:
Oh, good catch! But it's probably still not needed.
`Nx.subtract(12, Nx.Constants.neg_infinity) |> Nx.abs`
gives
```
#Nx.Tensor<
  f32
  Inf
>
```
which is larger than any `tol`. :)

polvalente replied:
Yeah, I just don't _like_ comparisons and operations with inf. But even in the worst case where we get `lower_bound = neg_infinity`, `neg_infinity - neg_infinity = nan`, which will also fail your check because it breaks on `true`

krstopro replied:
The only way `lower_bound` ends up being `neg_infinity` at that point is due to numerical error as it's updated `lower_bound = log_prob_norm` at line 169. But yeah, let's avoid operations with infinity.

polvalente reviewed (commented):
Lots of really minor comments. Overally this is looking great!

krstopro commented:
@msluszniak @polvalente Thanks for the support and given feedback! I've addressed most of the suggestions provided; see the comments above (except for vectorization that I would leave for now). There are few more remarks that I would like to make; see the comments below.

msluszniak commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,418 @@
+  end
+
+  defnp e_step(x, weights, means, precisions_cholesky) do
+    {log_prob_norm, log_resp} = estimate_log_prob_resp(x, weights, means, precisions_cholesky)
+    {Nx.mean(log_prob_norm), log_resp}
```
```suggestion
    {log_prob_norm, log_responsibilities} = estimate_log_prob_responsibilities(x, weights, means, precisions_cholesky)
    {Nx.mean(log_prob_norm), log_responsibilities}
```

msluszniak commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,418 @@
+    precisions_cholesky
+  end
+
+  defnp estimate_log_prob_resp(x, weights, means, precisions_cholesky) do
+    weighted_log_prob = estimate_weighted_log_prob(x, weights, means, precisions_cholesky)
+    log_prob_norm = logsumexp(weighted_log_prob)
+    log_resp = weighted_log_prob - Nx.new_axis(log_prob_norm, 1)
+    {log_prob_norm, log_resp}
+  end
```
```suggestion
  defnp estimate_log_prob_responsibilities(x, weights, means, precisions_cholesky) do
    weighted_log_prob = estimate_weighted_log_prob(x, weights, means, precisions_cholesky)
    log_prob_norm = logsumexp(weighted_log_prob)
    log_responsibilities = weighted_log_prob - Nx.new_axis(log_prob_norm, 1)
    {log_prob_norm, log_responsibilities}
  end
```

msluszniak commented:
Are there any blockers for this PR?

krstopro commented:
@msluszniak Don't think so. I just need to address few suggestions that you made. Could you have a look at the two reviews I opened? The `logsumexp` one might be important.

krstopro commented:
> Sure, but tbh I cannot see them 😅

@msluszniak They are probably collapsed among many others :)
Tagged you in the comment.

msluszniak commented:
Are you sure that your comments are not in `pending` mode or in the batch and not pushed or sth? Because I really cannot find anything

krstopro commented:
> Are you sure that your comments are not in `pending` mode or in the batch and not pushed or sth? Because I really cannot find anything

Sorry, they were in pending mode and I don't think I am able to submit them. Hence, I'll write them here.
Line 357-366 includes implementation to `logsumexp` which basically calculates the logarithm of the sum of the exponentials of the tensor in a numerically stable way. This is a well known trick that is implemented in numerical libraries, e.g. see SciPy
https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.logsumexp.html.
However, I am not aware that such a function exists in `Nx`. Is there one or perhaps I should raise an issue about implementing it?

Line 292: doing {num_gaussians, num_features, num_features} = Nx.shape(covariances) inside `defn` won't work, even though it works outside. I suppose this is intended?

msluszniak commented:
I'm almost sure that there is no such function implemented in `Nx` so feel free to send a PR with it. I'm not sure if we eventually want to have this type of functions in `Nx` or in `Scholar`, but it will be helpful.

This line won't work in `defn` because you're doing pattern matching. If you want to assert that the shape is as follows you need to check for it separately.

polvalente commented:
You can use `case` for checking the validity of the shape inside defn

krstopro commented:
> You can use `case` for checking the validity of the shape inside defn

I see, thanks. Here it is not an issue, just wanted to check that this is intended.
Anyway, if everything else is fine can you approve the PR?

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,433 @@
+  expectation of the Gaussian assignment for each data point x and the M-step which updates the
+  parameters to maximize the expectations found in E-step. While every iteration of the algorithm
+  is guaranteed to improve the log-likelihood, the final result depends on the initial values of
+  the parameters and the entire procedure should be repeated several times.
```
```suggestion
  the parameters. Thus, the algorithm consists of repeating this procedure several times to find the best result. 
```

WDYT? @msluszniak @krstopro

krstopro replied:
I would go for "Thus the procedure consists of repeating the algorithm several times and taking the best obtained result" as "algorithm" refers to EM algorithm mentioned before.

josevalim replied:
```suggestion
  the parameters. Thus the procedure consists of repeating the algorithm several times
  and taking the best obtained result.
```

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,433 @@
+
+  defnp estimate_gaussian_parameters(x, responsibilities, covariance_regularization_eps) do
+    nk = Nx.sum(responsibilities, axes: [0])
+    means = Nx.dot(Nx.transpose(responsibilities), x / nk)
```
```suggestion
    means = Nx.dot(responsibilities, [0], x / nk, [0])
```

I believe this will achieve the correct result too, but avoiding the transpose.

```elixir
iex(37)> k = Nx.Random.key(System.system_time())
#Nx.Tensor<
  u32[2]
  EXLA.Backend<host:0, 0.3654900815.2985164818.202632>
  [393393791, 2867920131]
>
iex(38)> {t1, k} = Nx.Random.uniform(k, shape: {3, 4})
{#Nx.Tensor<
   f32[3][4]
   EXLA.Backend<host:0, 0.3654900815.2985164818.203175>
   [
     [0.13576042652130127, 0.8412826061248779, 0.4484595060348511, 0.5624940395355225],
     [0.3512943983078003, 0.004010677337646484, 0.32932519912719727, 0.3748883008956909],
     [0.5318529605865479, 0.8757333755493164, 0.39265263080596924, 0.1744011640548706]
   ]
 >,
 #Nx.Tensor<
   u32[2]
   EXLA.Backend<host:0, 0.3654900815.2985164818.203177>
   [1882076740, 466335474]
 >}
iex(39)> {t2, k} = Nx.Random.uniform(k, shape: {3, 5})
{#Nx.Tensor<
   f32[3][5]
   EXLA.Backend<host:0, 0.3654900815.2985164818.203698>
   [
     [0.36859428882598877, 0.4500901699066162, 0.6306818723678589, 0.5150032043457031, 0.25934934616088867],
     [0.5741105079650879, 0.45720410346984863, 0.05595088005065918, 0.6362648010253906, 0.32299208641052246],
     [0.1018531322479248, 0.21718335151672363, 0.5860718488693237, 0.7845090627670288, 0.2448132038116455]
   ]
 >,
 #Nx.Tensor<
   u32[2]
   EXLA.Backend<host:0, 0.3654900815.2985164818.203700>
   [1307000727, 4256395817]
 >}
iex(40)> Nx.equal(Nx.dot(Nx.transpose(t1), t2), Nx.dot(t1, [-2], t2, [-2]))       #Nx.Tensor<
  u8[4][5]
  EXLA.Backend<host:0, 0.3654900815.2985164818.203708>
  [
    [1, 1, 1, 1, 1],
    [1, 1, 1, 1, 1],
    [1, 1, 1, 1, 1],
    [1, 1, 1, 1, 1]
  ]
>
```

krstopro replied:
How about `means = Nx.dot(responsibilities, [0], x / nk, [0])`?
Using `[0]` instead of `[-2]` seems more understandable to me.

polvalente replied:
I went with -2 due to force of habit of supporting batched versions of algorithms haha

Both are fine to me if you're enforcing 2D tensors here

krstopro replied:
Yeah, I had a similar habit with Tensorflow / PyTorch. I would use `[0]` here as both are matrices.

polvalente commented on lib/scholar/cluster/gmm.ex:
```diff
@@ -0,0 +1,433 @@
+            },
+            k < num_gaussians do
+        diff = x - means[k]
+        covariance = Nx.dot(responsibilities[[.., k]] * Nx.transpose(diff), diff) / nk[k]
```
```suggestion
        covariance = Nx.dot(responsibilities[[.., k]] * diff, [-2], diff / nk[k], [-2])
```

krstopro replied:
This won't work as `Nx.transpose` is only applied to `diff`, not the entire product. Adding another axis to `responsibilities[[.., k]]` solves the problem.
`covariance = Nx.dot(diff * Nx.new_axis(responsibilities[[.., k]], 1), [0], diff / nk[k], [0])`

josevalim commented:
@krstopro please verify @polvalente suggestions above and we can ship it :)

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
